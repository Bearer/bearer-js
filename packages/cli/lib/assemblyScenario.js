const serviceClient = require('./serviceClient')

module.exports = (
  scenarioTitle,
  emitter,
  {
    DeploymentUrl,
    bearerConfig: {
      authorization: {
        AuthenticationResult: { IdToken: token }
      }
    }
  }
) => {
  const deploymentClient = serviceClient(DeploymentUrl)
  emitter.emit('assemblyScenario:start')

  return deploymentClient
    .assemblyScenario(token, { bucketKey: scenarioTitle })
    .then(response => {
      if (response.statusCode === 201) {
        emitter.emit('assemblyScenario:success', response.body)
      } else {
        emitter.emit('assemblyScenario:failed', response)
      }
    })
    .catch(err => {
      emitter.emit('assemblyScenario:error', err)
    })
}
