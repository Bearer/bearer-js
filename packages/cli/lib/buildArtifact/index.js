const archiver = require('archiver')
const globby = require('globby')
const { promisify } = require('util')
const fs = require('fs')
const uuidv4 = require('uuid/v4')
const readFileAsync = promisify(fs.readFile)
const pathJs = require('path')
const webpack = require('webpack')

const prepareConfig = require('./prepareConfig')
const attachConfig = require('./attachConfig')
const addFilesToArchive = require('./addFilesToArchive')
const generateHandler = require('./generateHandler')

const CONFIG_FILE = 'bearer.config.json'
const HANDLER_NAME = 'index.js'
const archive = archiver('zip', {
  zlib: { level: 9 }
})

module.exports = async (output, handler, { path, scenarioUuid }, emitter) => {
  output.on('close', () => {
    emitter.emit('buildArtifact:output:close', pathJs.resolve(output.path))
  })

  output.on('end', () => {
    emitter.emit('buildArtifact:output:end')
  })

  archive.on('warning', err => {
    if (err.code === 'ENOENT') {
      emitter.emit('buildArtifact:archive:warning:ENOENT', err)
    } else {
      throw err
    }
  })

  archive.pipe(output)

  emitter.emit('buildArtifact:start', { scenarioUuid })
  // generate javascript files
  // copy package.json
  // create config
  // zip

  await transpileIntents(path)

  await prepareConfig(path, scenarioUuid)
    .then(async config => {
      emitter.emit('buildArtifact:configured', { intents: config.intents })

      await attachConfig(archive, JSON.stringify(config, null, 2), {
        name: CONFIG_FILE
      })

      archive.append(generateHandler(config), { name: HANDLER_NAME })
      await addFilesToArchive(archive, path)
    })
    .then(() => {
      archive.finalize()
    })
    .catch(console.error)

  return new Promise((resolve, reject) => {
    output.on('close', () => {
      resolve(archive)
    })

    archive.on('error', reject)
  })
}

function transpileIntents(path) {
  return new Promise((resolve, reject) => {
    globby([`${path}/*.ts`]).then(files => {
      const entries = files.reduceRight(
        (entriesAcc, file) => ({
          ...entriesAcc,
          [pathJs.basename(file).split('.')[0]]: file
        }),
        {}
      )
      webpack(
        {
          mode: 'production',
          entry: entries,
          module: {
            rules: [
              {
                test: /\.tsx?$/,
                loader: 'ts-loader',
                exclude: /node_modules/,
                options: {
                  compilerOptions: {
                    allowUnreachableCode: false,
                    declaration: false,
                    lib: ['es2017'],
                    noUnusedLocals: false,
                    noUnusedParameters: false,
                    allowSyntheticDefaultImports: true,
                    experimentalDecorators: true,
                    moduleResolution: 'node',
                    module: 'es6',
                    target: 'es2017'
                  }
                }
              }
            ]
          },
          resolve: {
            extensions: ['.tsx', '.ts', '.js']
          },
          target: 'node',
          output: {
            libraryTarget: 'commonjs2',
            filename: '[name].js',
            path: pathJs.join(pathJs.resolve(path), 'dist')
          },
          context: pathJs.resolve(path)
        },
        (err, stats) => {
          if (err || stats.hasErrors()) {
            console.log('[BEARER]', 'stats', stats.toJson('verbose'))
            reject(false)
          } else {
            resolve(true)
          }
        }
      )
    })
  })
}
