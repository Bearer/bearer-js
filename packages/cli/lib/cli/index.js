class CLI {
  constructor(program, emitter, config) {
    this.program = program
    this.emitter = emitter
    this.config = config
  }

  parse(argv) {
    this.program.parse(argv)

    // Display help if no command specified
    if (!argv.slice(2).length) {
      this.program.outputHelp()
      return
    }
  }

  use(cliCommand) {
    cliCommand.useWith(this.program, this.emitter, this.config)
  }
}

module.exports = { CLI }
