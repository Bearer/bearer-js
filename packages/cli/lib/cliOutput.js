const term = require('terminal-kit').terminal

module.exports = emitter => {
  emitter.on('buildArtifact:output:close', output_path => {
    term.white('Bearer: ')
    term.yellow('Artifact stored in ')
    term(output_path)
    term('\n')
  })

  emitter.on('buildArtifact:output:end', () => {
    console.log('Data has been drained')
  })

  emitter.on('buildArtifact:archive:warning:ENOENT', err => {
    term.red(`Warning ${err}`)
  })

  emitter.on('buildArtifact:start', ({ scenarioUuid }) => {
    term.white('Bearer: ')
    term.yellow(`Building scenario package artifact... [${scenarioUuid}]`)
    term('\n')
  })

  emitter.on('buildArtifact:configured', ({ intents }) => {
    const intentNames = intents.map(intent => Object.keys(intent))
    term.white('Bearer: ')
    term.yellow(`Artefact configured : ${intentNames.join(' | ')}`)
    term('\n')
  })
  emitter.on('pushScenario:unauthorized', ({ message }) => {
    term.white('Bearer: ')
    term.red(`ERROR: ${message}`)
    term('\n')
    term.white('Bearer: ')
    term.red(`Please try to `)
    term('bearer login ')
    term.red('again.')
    term('\n')
  })

  emitter.on('pushScenario:httpError', res => {
    term.white('Bearer: ')
    term.red(`ERROR: ${JSON.stringify(res, null, 2)}`)
    term('\n')
  })

  emitter.on('pushScenario:error', error => {
    term.white('Bearer: ')
    term.red(`There was an error when trying to push scenario: `)
    term('\n')
    term(error.toString())
    term('\n')
  })

  emitter.on('pushScenario:start', Key => {
    term.white('Bearer: ')
    term.yellow(`Pushing scenario ${Key}...`)
    term('\n')
  })

  emitter.on('pushScenario:uploadPackage:error', (err, packagePath) => {
    term.red(
      `There was an error when trying to push the package: ${packagePath}\n`
    )
    term.white(`Error: ${err}\n`)
  })

  emitter.on('pushScenario:uploadPackage:success', () => {
    term.white('Bearer: ')
    term.yellow('Scenario has been uploaded.')
    term('\n')
  })

  emitter.on('assemblyScenario:start', () => {
    term.white('Bearer: ')
    term.yellow('Building intents...')
    term('\n')
  })

  emitter.on('assemblyScenario:success', body => {
    term.white('Bearer: ')
    term.yellow('Intents created.')
    term('\n')

    term.white('PackageManager: ')
    term.yellow(JSON.stringify(body))
    term('\n')
  })

  emitter.on('assemblyScenario:failed', response => {
    term.white('Bearer: ')
    term.yellow('Something went wrong...')
    term(JSON.stringify(response, null, 2))
    term('\n')
  })

  emitter.on('assemblyScenario:error', err => {
    term.white('Bearer: ')
    term.red(`There was an error while trying to start the assembly: ${err}`)
    term('\n')
  })

  emitter.on('generateIntent:fileGenerated', path => {
    term.white('Bearer: ')
    term.yellow(`Bootstrapped a file: ${path}`)
    term('\n')
  })
  emitter.on('signUp:userCreated', email => {
    term.white('Bearer: ')
    term.yellow('User created: ')
    term.white(email)
    term('\n')
  })

  emitter.on('credentialsUpdated', configPath => {
    term.white('Bearer: ')
    term.yellow('Credentials and configuration stored in: ')
    term.white(configPath)
    term('\n')
  })
  emitter.on('signin:authenticateUser:getUserAttirbutes:failed', err => {
    term.white('Bearer: ')
    term.red(`There was en error while trying to fetch user attributes: ${err}`)
    term('\n')
  })

  emitter.on('signin:authenticateUser:failed', err => {
    term.white('Bearer: ')
    term.red(
      `There was en error while trying to fetch authenticate user: ${JSON.stringify(
        err
      )}`
    )
    term('\n')
  })

  emitter.on('signUp:error', err => {
    term.white('Bearer: ')
    term.red(`There was en error while trying to fetch sign up an user: ${err}`)
    term('\n')
  })

  emitter.on('developerIdFound', devId => {
    term.white('Bearer: ')
    term.red(`Your developerId: ${devId}`)
    term('\n')
  })

  emitter.on('scenarioTitle:missing', devId => {
    term.white('Bearer: ')
    term.red('Missing scenarioTitle')
    term('\n')
  })

  emitter.on('username:missing', () => {
    term.white('Bearer: ')
    term.red('Missing username.')
    term('\n')
    term.white('Bearer: ')
    term.yellow('Run ')
    term('bearer signup --email <email>')
    term.yellow(' first.')
    term('\n')
  })
  emitter.on('scenarioTitle:creationFailed', e => {
    term.white('Bearer: ')
    term.red("Couldn't store the scenarioTitle")
    term('\n')
    term(e)
    term('\n')
  })
  emitter.on('rootPath:doesntExist', () => {
    term.white('Bearer: ')
    term.red('Looks like you are not in scenario project directory.')
    term('\n')
    term.white('Bearer: ')
    term.red('Run ')
    term('bearer new <scenarioTitle>')
    term.red(' to bootstrap a new scenario.')
    term('\n')
  })

  emitter.on('intents:installingDependencies', () => {
    term.white('Bearer: ')
    term.yellow('Installing intents dependencies.')
    term('\n')
  })

  emitter.on('screens:installingDependencies', () => {
    term.white('Bearer: ')
    term.yellow('Installing screens dependencies.')
    term('\n')
  })

  emitter.on('screens:generateSetupComponent', () => {
    term.white('Bearer: ')
    term.yellow('Generating setup component.')
    term('\n')
  })

  emitter.on('screens:buildingDist', () => {
    term.white('Bearer: ')
    term.yellow('Building dist.')
    term('\n')
  })

  emitter.on('screens:pushingDist', () => {
    term.white('Bearer: ')
    term.yellow('Pushing dist.')
    term('\n')
  })

  emitter.on('screen:upload:start', () => {
    term.white('Bearer: ')
    term.yellow('Uploading screens...')
    term('\n')
  })

  emitter.on('screen:upload:success', () => {
    term.white('Bearer: ')
    term.yellow('Screens uploaded successfully.')
    term('\n')
  })

  emitter.on('screen:upload:error', e => {
    term.white('Bearer: ')
    term.red('ERROR: Screens upload failed.')
    term('\n')
    term(e.toString())
    term('\n')
  })

  emitter.on('screen:fileUpload:error', e => {
    term.white('Bearer: ')
    term.red('ERROR: Screen file upload failed.')
    term('\n')
    term(e.toString())
    term('\n')
  })
  emitter.on('screen:fileUpload:success', distPath => {
    term(distPath)
    term('\n')
  })

  emitter.on('screen:fileUpload:failure', distPath => {
    term.white('Bearer: ')
    term.red("Couldn't upload a file")
    term('\n')
    term(distPath)
    term('\n')
  })

  emitter.on('storeCredentials:success', referenceId => {
    term.white('Bearer: ')
    term.yellow('referenceId: ')
    term(referenceId)
    term('\n')
  })

  emitter.on('storeCredentials:failure', e => {
    term.white('Bearer: ')
    term.red('There was an error while trying to save credentials')
    term('\n')
    term(e.toString())
    term('\n')
  })

  emitter.on('signup:success', body => {
    term.white('Bearer: ')
    term.yellow('successfully signed up to bearer.')
    term('\n')
    term.white('Bearer: ')
    term(JSON.stringify(body))
    term.yellow(' saved to ')
    term('~/.bearerrc')
    term('\n')
  })

  emitter.on('signup:failure', body => {
    term.white('Bearer: ')
    term.red('There was an error while trying to signup to bearer')
    term('\n')
    term.white('IntegrationService: ')
    term.red(body.message)
    term('\n')
  })

  emitter.on('signup:error', e => {
    term.white('Bearer: ')
    term.red('There was an error while trying to signup to bearer')
    term('\n')
    term.white('Error: ')
    term.red(e.toString())
    term('\n')
  })

  emitter.on('login:success', body => {
    term.white('Bearer: ')
    term.yellow('successfully logged in to bearer.')
    term('\n')
    term.white('Bearer: ')
    term.yellow('AccessToken saved to ')
    term('~/.bearerrc')
    term('\n')
  })

  emitter.on('login:userFound', Username => {
    term.white('Bearer: ')
    term.yellow('Hello ')
    term(Username)
    term.yellow('!')
    term('\n')
  })
  emitter.on('login:failure', ({ message }) => {
    term.white('Bearer: ')
    term.red('There was an error while trying to login to bearer')
    term('\n')
    term.white('IntegrationService: ')
    term.red(message)
    term('\n')
  })

  emitter.on('login:error', ({ body: { message } }) => {
    term.white('Bearer: ')
    term.red('There was an error while trying to login to bearer')
    term('\n')
    term.white('Error: ')
    term.red(message)
    term('\n')
  })

  emitter.on('deploy:started', () => {
    term.white('Bearer: ')
    term.yellow('Starting scenario deployment')
    term('\n')
  })

  emitter.on('deploy:finished', ({ setupUrl }) => {
    term.white('Bearer: ')
    term('\n')
    term.yellow(`Scenario setup: `)
    term.white(setupUrl)
    term('\n')
  })

  emitter.on('invalidateCloudFront:success', () => {
    term.white('Bearer: ')
    term.yellow('Screen invalidation success.')
    term('\n')
  })

  emitter.on('invalidateCloudFront:invalidationFailed', ({ message }) => {
    term.white('Bearer: ')
    term.red("Couldn't invalidate screens cache.")
    term('\n')
    term.white('Error: ')
    term.red(message)
    term('\n')
  })

  emitter.on('invalidateCloudFront:error', ({ message }) => {
    term.white('Bearer: ')
    term.red('There was an error while trying to invalidate screens cache.')
    term('\n')
    term.white('Error: ')
    term.red(message)
    term('\n')
  })

  /* ********* Start output ********* */

  emitter.on('start:generate:buildFolder', () => {
    term.white('Bearer: ')
    term.yellow('Generating .build folder ')
    term('\n')
  })

  emitter.on('start:generate:stencilConfig', () => {
    term.white('Bearer: ')
    term.yellow('Generating stencil configuration')
    term('\n')
  })

  emitter.on('start:symlinkNodeModules', () => {
    term.white('Bearer: ')
    term.yellow('Symlinking node_modules')
    term('\n')
  })

  emitter.on('start:symlinkPackage', () => {
    term.white('Bearer: ')
    term.yellow('Symlinking package.json')
    term('\n')
  })

  emitter.on('start:watchers', () => {
    term.white('Bearer: ')
    term.yellow('Starting watchers')
    term('\n')
  })

  emitter.on('start:watchers:stencil:stdout', ({ data }) => {
    term.white('Bearer: ')
    term.yellow('[watcher-stencil] ')
    term.green(data)
  })

  emitter.on('start:watchers:stencil:stderr', ({ data }) => {
    term.white('Bearer: ')
    term.yellow('[watcher-stencil] ')
    term.green(data)
  })

  // emitter.emit('start:watchers:stencil:stdout', )

  emitter.on('start:failed', ({ error }) => {
    term.white('Bearer: ')
    term.red('An error occured')
    term('\n')
    term.white('    Error: ')
    term.red(error)
    term('\n')
  })
}
