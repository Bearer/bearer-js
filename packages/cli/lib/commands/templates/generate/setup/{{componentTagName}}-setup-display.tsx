/*
  The purpose of this component is to save scenario credentials.
*/

import { Component } from "@bearer/core";
import "@bearer/ui";

@Component({
  tag: "{{componentTagName}}-setup-display",
  shadow: true
})
export class {{scenarioTitle}}SetupDisplay {
  render() {
    return (
      <bearer-setup-display scenarioId="BEARER_SCENARIO_ID" />
    )
  }
}

