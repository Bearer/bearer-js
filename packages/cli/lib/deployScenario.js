const pathJs = require('path')
const fs = require('fs')
const util = require('util')
const exec = util.promisify(require('child_process').exec)
const { prepare, transpileStep } = require('./commands/startCommand')
const buildArtifact = require('./buildArtifact')
const pushScenario = require('./pushScenario')
const pushScreens = require('./pushScreens')
const assembly = require('./assemblyScenario')
const refreshToken = require('./refreshToken')
const invalidateCloudFront = require('./invalidateCloudFront')
const developerPortal = require('./developerPortal')

function buildIntents(rootLevel, scenarioUuid, emitter, config) {
  return new Promise(async (resolve, reject) => {
    const artifactDirectory = pathJs.join(rootLevel, '.bearer')
    const intentsDirectory = pathJs.join(rootLevel, 'intents')

    if (!fs.existsSync(artifactDirectory)) {
      fs.mkdirSync(artifactDirectory)
    }
    try {
      const scenarioArtifact = pathJs.join(
        artifactDirectory,
        `${scenarioUuid}.zip`
      )
      const handler = pathJs.join(artifactDirectory, config.HandlerBase)
      const output = fs.createWriteStream(scenarioArtifact)

      emitter.emit('intents:installingDependencies')
      await exec('yarn install', { cwd: intentsDirectory })

      await buildArtifact(
        output,
        handler,
        { path: intentsDirectory, scenarioUuid },
        emitter
      )
      return resolve(scenarioArtifact)
    } catch (e) {
      return reject(e)
    }
  })
}

const deployIntents = ({ scenarioUuid }, emitter, config) =>
  new Promise(async (resolve, reject) => {
    const { rootPathRc } = config

    if (!rootPathRc) {
      emitter.emit('rootPath:doesntExist')
      process.exit(1)
    }

    const rootLevel = pathJs.dirname(rootPathRc)

    try {
      const scenarioArtifact = await buildIntents(
        rootLevel,
        scenarioUuid,
        emitter,
        config
      )
      await pushScenario(
        scenarioArtifact,
        {
          Key: scenarioUuid
        },
        emitter,
        config
      )

      await assembly(scenarioUuid, emitter, config)
      resolve()
    } catch (e) {
      reject(e)
    }
  })

const deployScreens = ({ scenarioUuid }, emitter, config) =>
  new Promise(async (resolve, reject) => {
    const {
      scenarioConfig: { scenarioTitle },
      bearerConfig: { OrgId },
      BearerEnv
    } = config

    try {
      const { buildDirectory } = await prepare(emitter, config)()
      if (!buildDirectory) {
        process.exit(1)
        return false
      }

      await transpileStep(
        emitter,
        pathJs.join(buildDirectory, '..'),
        scenarioUuid,
        config.IntegrationServiceHost
      )

      emitter.emit('screens:generateSetupComponent')

      await exec('bearer generate --setup', { cwd: buildDirectory })

      emitter.emit('screens:generateConfigComponent')
      await exec('bearer generate --config', { cwd: buildDirectory })

      emitter.emit('screens:buildingDist')
      await exec('yarn build', {
        cwd: buildDirectory,
        pwd: buildDirectory,
        env: {
          BEARER_SCENARIO_ID: scenarioUuid,
          ...process.env,
          CDN_HOST: `https://static.${BearerEnv}.bearer.sh/${OrgId}/${scenarioTitle}/dist/${scenarioTitle}/`
        }
      })

      emitter.emit('screens:pushingDist')
      await pushScreens(buildDirectory, scenarioTitle, OrgId, emitter, config)

      emitter.emit('screen:upload:success')
      await invalidateCloudFront(emitter, config)
      return resolve()
    } catch (e) {
      emitter.emit('deployScenario:deployScreens:error', e)
      return reject(e)
    }
  })

module.exports = {
  deployIntents,
  deployScreens,
  buildIntents,
  deployScenario: ({ scenarioUuid }, emitter, config) =>
    new Promise(async (resolve, reject) => {
      let calculatedConfig = config

      try {
        const { ExpiresAt } = config.bearerConfig

        if (ExpiresAt < Date.now()) {
          calculatedConfig = await refreshToken(config, emitter)
        }
        await developerPortal(emitter, 'predeploy', calculatedConfig)
        await deployIntents({ scenarioUuid }, emitter, calculatedConfig)
        await deployScreens({ scenarioUuid }, emitter, calculatedConfig)
        await developerPortal(emitter, 'deployed', calculatedConfig)
        resolve()
      } catch (e) {
        await developerPortal(emitter, 'cancelled', calculatedConfig)
        reject(e)
      }
    })
}
