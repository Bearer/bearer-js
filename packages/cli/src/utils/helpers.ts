import * as copy from 'copy-template-dir'
import * as path from 'path'

import BaseCommand from '../BaseCommand'

export function copyFiles(
  command: BaseCommand,
  sourceDirectory: string,
  outPutDirectory: string,
  vars: any
): Promise<boolean | string> {
  const base = process.cwd()
  function printFiles(files: Array<string>) {
    files.forEach(file => {
      command.log(command.colors.gray(`    create: `) + command.colors.white(file.replace(base + '/', '')))
    })
  }

  return new Promise((resolve, reject) => {
    copy(path.join(__dirname, '..', '..', 'templates', sourceDirectory), outPutDirectory, vars, (err, createdFiles) => {
      if (err) {
        reject(err)
      } else {
        printFiles(createdFiles)
        resolve(true)
      }
    })
  })
}
