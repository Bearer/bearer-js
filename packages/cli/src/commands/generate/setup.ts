import { flags } from '@oclif/command'
import * as fs from 'fs'
import * as path from 'path'

import BaseCommand from '../../BaseCommand'
import { RequireScenarioFolder } from '../../utils/decorators'
import { copyFiles } from '../../utils/helpers'

export default class GenerateSetup extends BaseCommand {
  static description = 'Generate a Bearer Setup'
  static hidden = true
  static flags = {
    ...BaseCommand.flags,
    help: flags.help({ char: 'h' }),
    force: flags.boolean({})
  }

  static args = []

  @RequireScenarioFolder()
  async run() {
    const { flags } = this.parse(GenerateSetup)
    if (flags.force || !setupExists(this.locator.srcViewsDir)) {
      const fields = this.scenarioAuthConfig.setupViews
      if (fields && fields.length) {
        try {
          await copyFiles(
            this,
            'generate/setup',
            this.locator.srcViewsDir,
            this.getVars(this.bearerConfig.scenarioConfig.scenarioTitle, fields)
          )
          this.success('Setup components successfully generated! ðŸŽ‰')
        } catch (e) {
          this.error(e)
        }
      } else {
        this.warn(`No setupViews key found within auth.config.json file. skipping`)
      }
    } else {
      this.warn('Setup files already exists, use --force flag to overwrite existing ones.')
    }
  }

  getVars(scenarioTitle: string, fields: any) {
    return {
      componentName: this.case.pascal(scenarioTitle),
      componentTagName: this.case.kebab(scenarioTitle),
      fields: JSON.stringify(fields)
    }
  }
}

// Note: using Or condition in case the developer delete one but customized the other component
function setupExists(location: string): boolean {
  return (
    fs.existsSync(path.join(location, 'setup-action.tsx')) || fs.existsSync(path.join(location, 'setup-display.tsx'))
  )
}
