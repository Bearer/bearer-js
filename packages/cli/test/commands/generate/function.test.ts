import GenerateFunction from '../../../src/commands/generate/function'
import { ensureBearerStructure } from '../../helpers/setup'
import { readFile } from '../../helpers/utils'
import { Authentications } from '@bearer/types/lib/authentications'

describe('Generate', () => {
  let bearerPath: string
  let result: string[]

  describe.each(Object.values(Authentications))(`%s - generate:function`, authType => {
    beforeAll(() => {
      result = []
      jest.spyOn(process.stdout, 'write').mockImplementation(val => {
        result.push(val)
        return true
      })
      bearerPath = ensureBearerStructure({ authConfig: { authType }, folderName: authType })
    })

    it('Fetch function', async () => {
      await GenerateFunction.run(['FetchDataFunction', '--path', bearerPath])
      expect(result.join()).toContain('Function generated')
      expect(readFile(bearerPath, 'functions', 'FetchDataFunction.ts')).toMatchSnapshot()
    })
  })
})
