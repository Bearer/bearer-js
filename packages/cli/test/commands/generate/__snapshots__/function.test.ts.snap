// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Generate APIKEY - generate:function Fetch function 1`] = `
"import { TAPIKEYAuthContext, FetchData, TFetchActionEvent, TFetchPromise } from '@bearer/functions'
// Uncomment the line below to use the API Client
// import Client from './client'

export default class FetchDataFunctionFunction extends FetchData implements FetchData<ReturnedData, any, TAPIKEYAuthContext> {
  async action(event: TFetchActionEvent<Params, TAPIKEYAuthContext>): TFetchPromise<ReturnedData> {
    // const token = event.context.authAccess.apiKey
    // Put your logic here
    return { data: [] }
  }

  // Uncomment the line below to restrict the function to be called only from a server-side context
  // static serverSideRestricted = true

}

/**
 * Typing
 */
export type Params = {
  // name: string
}

export type ReturnedData = {
  // foo: string[]
}
"
`;

exports[`Generate BASIC - generate:function Fetch function 1`] = `
"import { TBASICAuthContext, FetchData, TFetchActionEvent, TFetchPromise } from '@bearer/functions'
// Uncomment the line below to use the API Client
// import Client from './client'

export default class FetchDataFunctionFunction extends FetchData implements FetchData<ReturnedData, any, TBASICAuthContext>{
  async action(event: TFetchActionEvent<Params, TBASICAuthContext>): TFetchPromise<ReturnedData> {
    // const { username, password } = event.context.authAccess
    // Put your logic here
    return { data: [] }
  }

  // Uncomment the line below to restrict the function to be called only from a server-side context
  // static serverSideRestricted = true

}

/**
 * Typing
 */
export type Params = {
  // name: string
}

export type ReturnedData = {
  // foo: string[]
}
"
`;

exports[`Generate CUSTOM - generate:function Fetch function 1`] = `
"import { TCUSTOMAuthContext, FetchData, TFetchActionEvent, TFetchPromise } from '@bearer/functions'
// Uncomment the line below to use the API Client
// import Client from './client'

export default class FetchDataFunctionFunction extends FetchData implements FetchData<ReturnedData, any, TCUSTOMAuthContext> {
  async action(event: TFetchActionEvent<Params, TCUSTOMAuthContext>): TFetchPromise<ReturnedData> {
    // Put your logic here
    return { data: [] }
  }

  // Uncomment the line below to restrict the function to be called only from a server-side context
  // static serverSideRestricted = true

}

/**
 * Typing
 */
export type Params = {
  // name: string
}

export type ReturnedData = {
  // foo: string[]
}
"
`;

exports[`Generate NONE - generate:function Fetch function 1`] = `
"import { FetchData, TFetchActionEvent, TFetchPromise } from '@bearer/functions'
// Uncomment the line below to use the API Client
// import Client from './client'

export default class FetchDataFunctionFunction extends FetchData implements FetchData<ReturnedData, any> {
  async action(event: TFetchActionEvent<Params>): TFetchPromise<ReturnedData> {
    // Put your logic here
    return { data: [] }
  }

  // Uncomment the line below to restrict the function to be called only from a server-side context
  // static serverSideRestricted = true

}

/**
 * Typing
 */
export type Params = {
  // name: string
}

export type ReturnedData = {
  // foo: string[]
}
"
`;

exports[`Generate OAUTH1 - generate:function Fetch function 1`] = `
"import { TOAUTH1AuthContext, FetchData, TFetchActionEvent, TFetchPromise } from '@bearer/functions'
// Uncomment the line below to use the API Client
// import Client from './client'

export default class FetchDataFunctionFunction extends FetchData implements FetchData<ReturnedData, any, TOAUTH1AuthContext> {
  async action(event: TFetchActionEvent<Params, TOAUTH1AuthContext>): TFetchPromise<ReturnedData> {
    // Put your logic here
    return { data: [] }
  }

  // Uncomment the line below to restrict the function to be called only from a server-side context
  // static serverSideRestricted = true

}

/**
 * Typing
 */
export type Params = {
  // name: string
}

export type ReturnedData = {
  // foo: string[]
}
"
`;

exports[`Generate OAUTH2 - generate:function Fetch function 1`] = `
"import { TOAUTH2AuthContext, FetchData, TFetchActionEvent, TFetchPromise } from '@bearer/functions'
// Uncomment the line below to use the API Client
// import Client from './client'

export default class FetchDataFunctionFunction extends FetchData implements FetchData<ReturnedData, any, TOAUTH2AuthContext> {
  async action(event: TFetchActionEvent<Params, TOAUTH2AuthContext>): TFetchPromise<ReturnedData> {
    // const token = event.context.authAccess.accessToken
    // Put your logic here
    return { data: [] }
  }

  // Uncomment the line below to restrict the function to be called only from a server-side context
  // static serverSideRestricted = true

}

/**
 * Typing
 */
export type Params = {
  // name: string
}

export type ReturnedData = {
  // foo: string[]
}
"
`;
